### PR TITLE
Add a structured list of solvers

### LINKED ISSUE #374: Add a structured list of solvers for different problem types
We have quite a lot of solvers now so I think it would be helpful to indicate which solvers can be used for which problem. I've started working on a list that could be included in docs. Please edit as you see fit.

The following list may help you select the right solver for your problem. Note that all solvers support Riemannian constraints.

* Generic nonlinear unconstrained:
  * Local derivative-free: Nelder-Mead
  * Local first order: Quasi-Newton, conjugate gradient descent, gradient descent
  * Local second order: adaptive regularization with cubics, trust regions
  * Global: particle swarm optimization
* Generic nonlinear constrained:
  * Local first order: augmented Lagrangian method, exact penalty method
* Convex unconstrained:
  * First order: subgradient method, convex bundle method
* Convex constrained:
  * First order: Frank-Wolfe method
* Special problems:
  * Nonlinear least squares: Levenberg-Marquardt (local first order)
  * Proximal: cyclic proximal point, Douglas-Rachford, Chambolle-Pock, primal-dual Riemannian semismooth Newton
  * Difference of convex (local first order)

### PR DESCRIPTION
**Pull request by kellertuer**
This is a start to resolve #374.

I am still not so sure how to best to the list.

* for now there is a rough classification based on “knowledge” required about the cost
* each solver gets a 1-2 sentence items block for now and (maybe in a newline or even with a new subsubsection) an example call to the solver. Both the name and the function should be linked.

Further ideas welcome – for now I just build the groups and 3 examples for solvers how I would introduce them in “super-short”.

### CODE REVIEW

File: docs/src/solvers/index.md [hunk 1]
```diff
@@ -1,43 +1,116 @@
...
 
 ```@meta
 CurrentModule = Manopt
 ```
 
-Solvers can be applied to [`AbstractManoptProblem`](@ref)s with solver
-specific [`AbstractManoptSolverState`](@ref).
+Optimisation problems can be classified with respect to several criteria.
+In the following we provide a grouping of the algorithms with respect to the “information”
+available for your optimisation problem
```
Comment (mateuszbaran) [inline docs/src/solvers/index.md:10]:
```suggestion
available about your optimisation problem
```

**Comment by mateuszbaran** [comment]
> * 
> * each solver gets a 1-2 sentence items block for now and (maybe in a newline or even with a new subsubsection) an example call to the solver. Both the name and the function should be linked.

I think we don't need an example call there, we can just link to an example that is available elsewhere. A couple of sentences about the solver is IMHO enough to provide an idea whether it's appropriate solver for a particular problems, and details should be elsewhere.

**Comment by kellertuer** [comment]
Ok, then without the call example.

**Comment by kellertuer** [comment]
I just finished the list. Currently, in every section, the algorithms are ordered alphabetically, we could order them also by “favourite ones first” in an oppinionated way.

I also kept an alphabetical table below, that now instead of name-function,state-“cost properties” has 3 columns name-function-state, where function refers to the high-level interface.

File: docs/src/solvers/index.md [hunk 2]
```diff
@@ -1,43 +1,114 @@
...
+Within the groups we provide short notes on advantages of the individual solvers,
+for example whether ``f`` should be convex, or nonconvex – or whether the method works
+locally or even globally.
+
+## Derivative Free
+
+For derivative free only function evaluations of ``f`` are used.
+
+* [Nelder-Mead](NelderMead.md) a simplex based variant, that is using ``d+1`` points, where ``d`` is the dimension of the manifold.
+* [Particle Swarm](particle_swarm.md) use the evolution of a set of points, called swarm, to explore the domain of the cost and find a minimizer.
```
Comment (mateuszbaran) [inline docs/src/solvers/index.md:25]:
```suggestion
* [Particle Swarm](particle_swarm.md) use the evolution of a set of points, called swarm, to explore the domain of the cost and find a minimizer.
* [CMA-ES](cma_es.md) uses a stochastic evolutionary strategy to perform minimization robust to local minima of the objective.
```

File: docs/src/solvers/index.md [hunk 3]
```diff
@@ -1,43 +1,114 @@
...
+* [Conjugate Gradient Descent](conjugate_gradient_descent.md) uses information from the previous descent direction to improve the current (gradient-based) one including several such update rules.
+* [Levenberg-Marquardt](LevenbergMarquardt.md) minimizes the square norm of ``f: \mathcal M→ℝ^d`` provided the gradients of the component functions, or in other words the Jacobian of ``f``.
+* The [Quasi-Newton Method](quasi_Newton.md) uses gradient evaluations to approximate the Hessian, which is then used in a Newton-like scheme, where both a limited memory and a full Hessian approximation are available with several different update rules.
+* [Steihaug-Toint Truncated Conjugate-Gradient Method](@ref tCG) a solver for a constrained problem defined on a tangent space.
+* [Stochastic Gradient Descent](stochastic_gradient_descent.md) is based on a splitting of ``f`` into several components ``f_i`` whose gradients are provided here. By randomly selecting these, steps towards (compnent-wise) descent directions are performed.
```
Comment (mateuszbaran) [inline docs/src/solvers/index.md:36]:
```suggestion
* [Stochastic Gradient Descent](stochastic_gradient_descent.md) is based on a splitting of ``f`` into a sum of several components ``f_i`` whose gradients are provided. Steps are performed according to gradients of randomly selected components.
```
I think ideally the interface should be such that the user provides a single function but it gets component number `i` as an argument.

Also, isn't this technically a splitting method?

File: docs/src/solvers/index.md [hunk 4]
```diff
@@ -1,43 +1,114 @@
...
+The following methods require the Riemannian subgradient ``∂f`` to be available.
+While the subgradient might be set-valued, the function should provide one of the subgradients.
+
+* The [Subgradient Method](subgradient.md) takes the negative subgradient as a step direction and can be combined with a step size.
+* The [Convex Bundle Method](convex_bundle_method.md) (CBM) uses a former collection of sub gradients at the previous iterates and iterate candidates to solve a local approximation to `f` in every iteration by solving a quadratic problem in the tangent space.
+* The [Proximal Bundle Method](proximal_bundle_method.md) works similar to CBM, but solves a proximal map-based problem in every iteration.
+
+## Second Order
+
+* [Adaptive Regularisation with Cubics](adaptive-regularization-with-cubics.md) locally builds a cubic model to determine a next descent direction
```
Comment (mateuszbaran) [inline docs/src/solvers/index.md:49]:
```suggestion
* [Adaptive Regularisation with Cubics](adaptive-regularization-with-cubics.md) locally builds a cubic model to determine the next descent direction.
```

File: docs/src/solvers/index.md [hunk 5]
```diff
@@ -1,43 +1,114 @@
...
+* The [Riemannian Trust-Regions Solver](trust_regions.md) builds a quadratic model within a trust region to determine a next descent direction
```
Comment (mateuszbaran) [inline docs/src/solvers/index.md:50]:
```suggestion
* The [Riemannian Trust-Regions Solver](trust_regions.md) builds a quadratic model within a trust region to determine the next descent direction.
```

**Comment by mateuszbaran** [comment]
> I just finished the list. Currently, in every section, the algorithms are ordered alphabetically, we could order them also by “favourite ones first” in an oppinionated way.

Maybe it would be sufficient to add a symbol that distinguishes algorithms known to be state-of-the-art for some problems (like quasi-Newton) and those that are here mostly for historical perspective and comparison (like gradient descent)? I don't think we can order algorithms from best to worst because that often depends on the problem being solved.

Anyway, I generally like the list and I think it's a very good addition :+1: .

File: docs/src/solvers/index.md (hunk 2 above)
Comment (kellertuer) [reply docs/src/solvers/index.md:25]: oh, did I forget that one? Sorry for that.

File: docs/src/solvers/index.md (hunk 3 above)
Comment (kellertuer) [reply docs/src/solvers/index.md:37]: Yes, technically both LM and SGD are splitting methods.

Comment (kellertuer) [reply docs/src/solvers/index.md]: Maybe one could rename the spitting section to “Nonsmooth Splitting methods”? That would be a remedy maybe that unifies what I collected there.

**Comment by kellertuer** [comment]
> Maybe it would be sufficient to add a symbol that distinguishes algorithms known to be state-of-the-art for some problems (like quasi-Newton) and those that are here mostly for historical perspective and comparison (like gradient descent)? I don't think we can order algorithms from best to worst because that often depends on the problem being solved.

For DocumenterInterlinks [I saw that their docs}(http://juliadocs.org/DocumenterInterLinks.jl/stable/syntax/) that they handed out gold and silver medals for recommended (and follow up) methods/syntax. We could do that as well and explain that somewhere in the beginning. We could add also a symbol for merely academic/teaching examples as well, but phrase that carefully, I would like to avoid saying “This alsogirthm is not recommended”. The only ones I would maybe not really recommend are some of the full-matrix-updates in QN.

**Comment by mateuszbaran** [comment]
I see, maybe that's actually unnecessary.

**Comment by kellertuer** [comment]
I think it might be nice to add, I am just not sure how to decide which ones to mark with a symbol then.

File: docs/src/solvers/index.md [hunk 6]
```diff
@@ -1,43 +1,115 @@
...
+* The [Subgradient Method](subgradient.md) takes the negative subgradient as a step direction and can be combined with a step size.
+* The [Convex Bundle Method](convex_bundle_method.md) (CBM) uses a former collection of sub gradients at the previous iterates and iterate candidates to solve a local approximation to `f` in every iteration by solving a quadratic problem in the tangent space.
+* The [Proximal Bundle Method](proximal_bundle_method.md) works similar to CBM, but solves a proximal map-based problem in every iteration.
+
+## Second Order
+
+* [Adaptive Regularisation with Cubics](adaptive-regularization-with-cubics.md) locally builds a cubic model to determine the next descent direction.
+* The [Riemannian Trust-Regions Solver](trust_regions.md) builds a quadratic model within a trust region to determine the next descent direction.
+
+## Splitting based
```
Comment (mateuszbaran) [inline docs/src/solvers/index.md:53]:
```suggestion
## Nonsmooth splitting based
```
Or maybe we can make two subsections, one with smooth splitting based (LM, SGD) and one with nonsmooth splitting-based.

Comment (kellertuer) [reply docs/src/solvers/index.md:53]: Well, to be precise the Alternating Gradient descent is also a splitting method then :)

Comment (mateuszbaran) [reply docs/src/solvers/index.md:53]: Sure, then let's add it there too :slightly_smiling_face:

**Comment by kellertuer** [comment]
I do like the current division into property based groups – I would like to think a bit about adding emojis to some of them to indicate certain recommendations (state-of-the-art, maybe something like fastest, or most used or so).

File: docs/src/solvers/index.md [hunk 7]
```diff
@@ -1,43 +1,124 @@
...
+| [Adaptive Regularisation with Cubics](adaptive-regularization-with-cubics.md) | [`adaptive_regularization_with_cubics`](@ref) | [`AdaptiveRegularizationState`](@ref) |
+| [Augmented Lagrangian Method](augmented_Lagrangian_method.md) | [`augmented_Lagrangian_method`](@ref) | [`AugmentedLagrangianMethodState`](@ref) |
+| [Chambolle-Pock](ChambollePock.md) | [`ChambollePock`](@ref) | [`ChambollePockState`](@ref) |
+| [Conjugate Gradient Descent](conjugate_gradient_descent.md) | [`conjugate_gradient_descent`](@ref) | [`ConjugateGradientDescentState`](@ref) |
+| [Convex Bundle Method](convex_bundle_method.md) | [`convex_bundle_method`](@ref) |  [`ConvexBundleMethodState`](@ref) |
+| [Cyclic Proximal Point](cyclic_proximal_point.md) | [`cyclic_proximal_point`](@ref) |  [`CyclicProximalPointState`](@ref) |
+| [Difference of Convex Algorithm](@ref solver-difference-of-convex) |
+[`difference_of_convex_algorithm`](@ref) | [`DifferenceOfConvexState`](@ref) |
```
Comment (mateuszbaran) [inline docs/src/solvers/index.md:103]:
```suggestion
| [Difference of Convex Algorithm](@ref solver-difference-of-convex) | [`difference_of_convex_algorithm`](@ref) | [`DifferenceOfConvexState`](@ref) |
```
Newline breaks the table a bit.

File: docs/src/solvers/index.md [hunk 8]
```diff
@@ -1,43 +1,124 @@
...
+
+### Nonsmooth
+
+If the gradient does not exist everywhere, that is if the splitting yields summands that are nonsmooth, usually methods based on proximal maps are used.
+
+* The [Chambolle-Pock](ChambollePock.md) algorithm uses a splitting ``f(p) = F(p) + G(Λ(p))``,
+  where ``G`` is defined on a manifold ``\mathcal N`` and we need the proximal map of its Fenchel dual. Both these functions can be non-smooth.
+* The [Cyclic Proximal Point](cyclic_proximal_point.md) 🫏 uses proximal maps of the functions from splitting ``f`` into summands ``f_i``
+* [Difference of Convex Algorithm](@ref solver-difference-of-convex) uses a splitting of the (nonconvex) function ``f = g - h`` into a difference of two functions; for each of these we require the gradient of ``g`` and the subgradient of ``h`` to state a sub problem in every iteration to be solved.
+* [Difference of Convex Proximal Point](@ref solver-difference-of-convex-proximal-point) | [`difference_of_convex_proximal_point`](@ref)
```
Comment (mateuszbaran) [inline docs/src/solvers/index.md:71]: Here we maybe should have a sentence about this solver?

Comment (kellertuer) [reply docs/src/solvers/index.md:71]: Oh, missed to write that. Will have to check that paper for short how to best formulate that. Thanks for spotting.

**Review by mateuszbaran — APPROVED** [review]

**Pull request merged**